src/encryption.rs

```diff
@@ -192,9 +192,8 @@ unsafe impl Send for DBEncryptionKeyManager {}
 unsafe impl Sync for DBEncryptionKeyManager {}
 
 impl DBEncryptionKeyManager {
-    pub fn new(key_manager: Arc<dyn EncryptionKeyManager>) -> DBEncryptionKeyManager {
-        // Size of Arc<dyn T>::into_raw is of 128-bits, which couldn't be used as C-style pointer.
-        // Boxing it to make a 64-bits pointer.
+    pub fn new(key_manager: Box<dyn EncryptionKeyManager>) -> DBEncryptionKeyManager {
```

Comment:
I think we can use generics to unbox further. The following can allow the caller to avoid boxing the type that implement `EncryptionKeyManager`. 
```
pub fn new<T: EncryptionKeyManager> (key_manager: T) -> DBEncryptionKeyManager {
        let ctx = Box::into_raw(Box::new(key_manager)) as *mut c_void;
        let instance = unsafe {
            crocksdb_ffi::crocksdb_encryption_key_manager_create(
                ctx,
                encryption_key_manager_destructor<T>,
                encryption_key_manager_get_file<T>,
                encryption_key_manager_new_file<T>,
                encryption_key_manager_delete_file<T>,
                encryption_key_manager_link_file<T>,
            )
        };
        DBEncryptionKeyManager { inner: instance }
}
```

Reply:
Sad it didn't work out. Essentially you need to preserve the `T` information to be able to restore a `void` pointer. With double boxing we can store it to vtable, but with generic constructor that information is lost.